pilot/pkg/proxy/envoy/config.go

```diff
@@ -515,20 +521,24 @@ func buildOutboundListeners(mesh *meshconfig.MeshConfig, sidecar model.Node, ins
 			operation = IngressTraceOperation
 		}
 
-		l := buildHTTPListener(mesh, sidecar, instances, routeConfig, WildcardAddress, port,
+		listener := buildHTTPListener(mesh, sidecar, instances, routeConfig, WildcardAddress, port,
 			fmt.Sprintf("%d", port), useRemoteAddress, operation, true, config)
-		listeners = append(listeners, l)
+		listeners = append(listeners, listener)
 		clusters = append(clusters, routeConfig.clusters()...)
 	}
 
 	return listeners, clusters
 }
 
+type buildClusterFunc func(hostname string, port *model.Port, labels model.Labels, isExternal bool) *Cluster
```

Comment:
This is nice. I think we should elevate this thought and have first class types for build cluster, build route, build listener, etc. And they should all get one big context struct that has all the info/handles it needs. cc @ZackButcher

Reply:
100% - I've been thinking about refactorings I want to do in this part of the codebase and high on my list is adding types throughout to provide more structure/better abstractions (the first new type being a context struct so we can get rid of like 4 parameters on every function :) ).